[NFC] Describe sycl-post-link support for dynamic linking (#14337)
intel/llvm: sycl/doc/design/SharedLibraries.md

### sycl-post-link changes

> In typical device code splitting, all dependent device code is brought into a split
> image in order to construct a complete image.  This is impossible with dynamic linking
> since the dynamic library may rely on functions that are not defined in the library.

Reviewer: ```suggestion
When device code splitting is performed during a typical compilation of a SYCL offloading program, all dependent device code is brought into a split
image in order to construct a complete image.  When device code splitting is performed during the compilation flow to generate a dynamic library, construction of a complete device image is not possible
since the dynamic library may rely on functions that are not defined in the library.
```

Author: Changed.

> Thus, when supporting dynamic linking a dependency to a canBeImportedFunction will not

Reviewer 2: ```suggestion
Thus, when supporting dynamic linking a dependency to a `canBeImportedFunction` will not
```

Reviewer: I would avoid the use of 'canBeImportedFunction' here. It makes the document a bit unreadable.

cause the dependent function to be added to the image.

> A canBeImportedFunction is:

Reviewer 2: ```suggestion
A `canBeImportedFunction` is:
```

Author: Changed.

  1. Not an intrinsic
  2. Name does not start with "__"
  3. Demangled name does not start with "__"

>   4. Must be a SYCL_EXTERNAL function

Reviewer 2: ```suggestion
  4. Must be a `SYCL_EXTERNAL` function
```

Author: Changed

In order to support dynamic linking of device code, `sycl-post-link` has

> 4 modifications:

Reviewer 2: ```suggestion
the following modifications:
```

Nit, just to avoid having to change the number if we ever decide to reword the list below somehow

Author: Changed

- Device images constructed to support dynamic libraries do not include

>   dependencies to canBeImportedFunctions.

Author: Changed.

> - Dependencies to a canBeImportedFunction are recorded in the imported symbols

Reviewer 2: ```suggestion
- Dependencies to a `canBeImportedFunction` are recorded in the imported symbols
```

Author: Changed

  property list.

> - An image that provides a canBeImportedFunction has the symbol recorded in the exported

Reviewer 2: ```suggestion
- An image that provides a `canBeImportedFunction` has the symbol recorded in the exported
```

Author: Changed.

  symbols property list.

> - All functions symbols that are not canBeImportedFunction are internalized

Reviewer 2: ```suggestion
- All functions symbols that are not `canBeImportedFunction` are internalized
```

Author: Changed.

> When generating a compile image for a kernel, the SYCL Run-Time will find the device

Reviewer 2: There is a dedicated section about RT changes, do we really need this whole paragraph here?

Author: Deleted.
